src/D2L.CodeStyle.Analyzers/Language/LikelyArgumentMismatchAnalyzer.cs

```diff
@@ -0,0 +1,669 @@
+		}
+
+		private static void RegisterAnalyzer( CompilationStartAnalysisContext context ) {
+			context.RegisterSyntaxNodeAction(
+				ctx => AnalyzeInvocation(
+					ctx,
+					(InvocationExpressionSyntax)ctx.Node
+				),
+				SyntaxKind.InvocationExpression
+			);
+		}
+
+		private static void AnalyzeInvocation(
+			SyntaxNodeAnalysisContext context,
+			InvocationExpressionSyntax expr
+		) {
+			if( expr.ArgumentList == null ) {
+				return;
+			}
+
+			if( expr.ArgumentList.Arguments.Count <= 1 ) {
+				return;
+			}
+
+			// get bipartite matching of the source
+			ImmutableArray<Edge> sourceMatching = MatchArguments(
+				context.SemanticModel,
+				expr.ArgumentList.Arguments
+			).ToImmutableArray();
+
+			if( sourceMatching.Length <= 1 ) {
+				return;
+			}
+
+			ImmutableArray<Edge> bestMatching = GenerateBestMatching( context, sourceMatching );
+
+			ImmutableDictionary<ArgumentDetails, Edge> sourceEdges = sourceMatching
+				.ToImmutableDictionary( x => x.Arg, x => x );
+			foreach( Edge edge in bestMatching ) {
+				Edge sourceEdge = sourceEdges[ edge.Arg ];
+				if( !sourceEdge.Equals( edge ) ) {
+					context.ReportDiagnostic( Diagnostic.Create(
+						Diagnostics.LikelyArgumentMismatch,
+						sourceEdge.Arg.Location,
+						sourceEdge.Arg.Name,
+						edge.Param.Name,
+						sourceEdge.Param.Name
+					) );
+				}
+			}
+		}
+
+		private static ImmutableArray<Edge> GenerateBestMatching(
+			SyntaxNodeAnalysisContext context,
+			ImmutableArray<Edge> sourceMatching
+		) {
+			ImmutableArray<ArgumentDetails> arguments = sourceMatching.Select( x => x.Arg ).ToImmutableArray();
+			ImmutableArray<ParameterDetails> parameters = sourceMatching.Select( x => x.Param ).ToImmutableArray();
+
+			int[,] costMatrix = GetCostMatrix( context.SemanticModel, arguments, parameters );
+
+			int[] assignments = HungarianAlgorithm.FindAssignments( costMatrix );
+			ImmutableArray<Edge> matching = ConvertToMatching(
+				assignments,
+				arguments,
+				parameters
+			);
+
+			return matching;
+		}
+
+		private static int[,] GetCostMatrix(
+			SemanticModel model,
+			ImmutableArray<ArgumentDetails> arguments,
+			ImmutableArray<ParameterDetails> parameters
+		) {
+			int[,] matrix = new int[ arguments.Length, parameters.Length ];
+
+			for( int i = 0; i < arguments.Length; ++i ) {
+				for( int j = 0; j < parameters.Length; ++j ) {
+					matrix[ i, j ] = GetCost(
+						model,
+						arguments[ i ],
+						parameters[ j ]
+					);
+				}
+			}
+
+			return matrix;
+		}
+
+		private static int GetCost(
+			SemanticModel model,
+			ArgumentDetails argument,
+			ParameterDetails parameter
+		) {
+			double cost = 10;
+			double costMultiplier = 1;
+
+			Conversion conversion = model.ClassifyConversion( argument.Syntax, parameter.Type );
+			if( !conversion.Exists ) {
+				// if the types don't match, make the cost large so we don't choose it
+				return int.MaxValue;
+			}
+
+			if( !conversion.IsIdentity && !conversion.IsUserDefined ) {

```

Comment:
What does it meant for `conversion.IsIdentity` to be `true`?

Reply:
Means no conversion. `long -> long`